File: src3/AdvisorInterface.java
import java.util.Scanner;

public class AdvisorInterface {

    private static final Scanner scanner = new Scanner(System.in);

    // Main method to run the Advisor Interface
    public void run() {
        while (true) {
            displayMenu();
            int choice = getUserInput();

            switch (choice) {
                case 1:
                    viewStudents();
                    break;
                case 2:
                    approveDisapproveCourses();
                    break;
                case 3:
                    viewNotifications();
                    break;
                case 4:
                    System.out.println("Logging out...");
                    return; // Logout
                default:
                    System.out.println("Invalid option. Please try again.");
            }
        }
    }

    // Display the Advisor menu
    private void displayMenu() {
        System.out.println("\n=== Advisor Menu ===");
        System.out.println("1. View Students");
        System.out.println("2. Approve/Disapprove Courses");
        System.out.println("3. View Notifications");
        System.out.println("4. Logout");
        System.out.print("Please select an option: ");
    }

    // Safely get user input as an integer
    private int getUserInput() {
        while (!scanner.hasNextInt()) {
            System.out.println("Invalid input. Please enter a number.");
            scanner.nextLine();
        }
        int choice = scanner.nextInt();
        scanner.nextLine(); // Clear buffer
        return choice;
    }

    // Method to view all students under this advisor
    private void viewStudents() {
        Advisor currentAdvisor = getCurrentAdvisor();

        if (currentAdvisor.getStudentList().isEmpty()) {
            System.out.println("No students are currently assigned to you.");
            return;
        }

        System.out.println("\n=== List of Students ===");
        currentAdvisor.viewStudentList();
    }

    // Method to approve or disapprove student courses
    private void approveDisapproveCourses() {
        Advisor currentAdvisor = getCurrentAdvisor();

        if (currentAdvisor.getAwaitingStudents().isEmpty()) {
            System.out.println("No students are awaiting approval.");
            return;
        }

        System.out.println("\n=== Awaiting Students ===");
        for (int i = 0; i < currentAdvisor.getAwaitingStudents().size(); i++) {
            Student student = currentAdvisor.getAwaitingStudents().get(i);
            System.out.println((i + 1) + ". " + student.getFirstName() + " " + student.getLastName());
        }

        System.out.print("Select a student (1 to " + currentAdvisor.getAwaitingStudents().size() + "): ");
        int studentIndex = getUserInput();
        
        if (studentIndex < 1 || studentIndex > currentAdvisor.getAwaitingStudents().size()) {
            System.out.println("Invalid selection.");
            return;
        }

        Student selectedStudent = currentAdvisor.getAwaitingStudents().get(studentIndex - 1);

        System.out.println("\n1. Approve Courses");
        System.out.println("2. Disapprove Courses");
        System.out.print("Select an option: ");
        int action = getUserInput();

        switch (action) {
            case 1:
                currentAdvisor.approveCourses();
                NotificationSystem.sendAdvisorNotification(currentAdvisor, selectedStudent.getFirstName() + " " + selectedStudent.getLastName() + "'s courses have been approved.");
                System.out.println("Courses approved for " + selectedStudent.getFirstName() + " " + selectedStudent.getLastName());
                break;
            case 2:
                currentAdvisor.disapproveCourses();
                NotificationSystem.sendAdvisorNotification(currentAdvisor, selectedStudent.getFirstName() + " " + selectedStudent.getLastName() + "'s courses have been disapproved.");
                System.out.println("Courses disapproved for " + selectedStudent.getFirstName() + " " + selectedStudent.getLastName());
                break;
            default:
                System.out.println("Invalid option.");
        }
    }

    // Method to view notifications for the advisor
    private void viewNotifications() {
        Advisor currentAdvisor = getCurrentAdvisor();

        NotificationSystem.viewAdvisorNotifications(currentAdvisor);

        if (currentAdvisor.getUnreadNotifications().isEmpty()) {
            System.out.println("You have no unread notifications.");
            return;
        }

        System.out.print("\nMark all notifications as read? (y/n): ");
        String choice = scanner.nextLine();

        if (choice.equalsIgnoreCase("y")) {
            NotificationSystem.markAdvisorNotificationsAsRead(currentAdvisor);
            System.out.println("All notifications marked as read.");
        } else {
            System.out.println("Notifications remain unread.");
        }
    }

    // Helper method to get the current logged-in advisor
    private Advisor getCurrentAdvisor() {
        return ObjectCreator.advisors.get(UserInterface.currentAdvisorsIndex);
    }
}

File: Iteration3/Python/AdvisorInterface.py
import NotificationSystem
import ObjectCreator
import UserInterface


class AdvisorInterface:

    def __init__(self):
        self.scanner = input  # Using Python's input function for user input.

    def run(self):

        while True:
            print("\n=== Advisor Menu ===")
            print("1. View Students")
            print("2. Approve/Disapprove Courses")
            print("3. View Notifications")
            print("4. Logout")
            try:
                choice = int(self.scanner("Please select an option: "))
            except ValueError:
                print("Invalid input. Please enter a number between 1 and 4.")
                continue

            if choice == 1:
                self.view_students()
            elif choice == 2:
                self.approve_disapprove_courses()
            elif choice == 3:
                self.view_notifications()
            elif choice == 4:
                print("Logging out...")
                break
            else:
                print("Invalid option. Please try again.")

    def view_students(self):

        current_advisor = self.get_current_advisor()
        if not current_advisor.get_student_list():
            print("No students assigned to you.")
        else:
            current_advisor.view_student_list()

    def approve_disapprove_courses(self):

        current_advisor = self.get_current_advisor()
        awaiting_students = current_advisor.get_awaiting_students()

        if not awaiting_students:
            print("No students awaiting approval.")
            return

        print("\nSelect a student to approve/disapprove courses:")
        for i, student in enumerate(awaiting_students, start=1):
            print(f"{i}. {student.get_first_name()} {student.get_last_name()}")

        try:
            student_index = int(self.scanner(f"Select student (1 to {len(awaiting_students)}): "))
        except ValueError:
            print("Invalid input. Please enter a number.")
            return

        if student_index < 1 or student_index > len(awaiting_students):
            print("Invalid selection.")
            return

        selected_student = awaiting_students[student_index - 1]
        current_advisor.set_adv_student(selected_student)

        print("\nSelect an option:")
        print("1. Approve Courses")
        print("2. Disapprove Courses")
        try:
            action = int(self.scanner("Select an option: "))
        except ValueError:
            print("Invalid input. Please enter 1 or 2.")
            return

        if action == 1:
            current_advisor.approve_courses()
            NotificationSystem.send_advisor_notification(
                current_advisor,
                f"Courses approved for {selected_student.get_first_name()} {selected_student.get_last_name()}."
            )
        elif action == 2:
            current_advisor.disapprove_courses()
            NotificationSystem.send_advisor_notification(
                current_advisor,
                f"Courses disapproved for {selected_student.get_first_name()} {selected_student.get_last_name()}."
            )
        else:
            print("Invalid option.")

    def view_notifications(self):

        current_advisor = self.get_current_advisor()
        NotificationSystem.view_advisor_notifications(current_advisor)

        choice = self.scanner("\nMark all notifications as read? (y/n): ")
        if choice.lower() == "y":
            NotificationSystem.mark_advisor_notifications_as_read(current_advisor)
            print("All notifications marked as read.")

    def get_current_advisor(self):

        try:
            return ObjectCreator.advisors[UserInterface.current_advisors_index]
        except IndexError:
            print("Error: No advisor is currently logged in.")
            return None
